### ISSUE TITLE
[Feature Request] Automatic switch to Offline Mode when disconnected from Wi-Fi

### ISSUE THREAD

**Issue by rigrig** [jmshrv/finamp#210]
It would be nice if Finamp could automatically switch to Offline Mode when disconnecting from Wi-Fi, to prevent mobile data usage. (And back to Online when reconnecting)

**Comment by UnicornsOnLSD** [comment]
This would be nice to have, I'd have to think about how to handle whether or not to automatically switch though (as someone may want to manually enable offline mode on WiFi, or vice-versa

**Comment by drazil100** [comment]
Ideally (and I'm pretty sure it works this way already but I'm not 100% certain) if the song is already downloaded it should play the downloaded copy rather than playing it from the server.

That said I think having this as an option users can toggle on to save bandwidth is a fantastic idea.

**Comment by UnicornsOnLSD** [comment]
> Ideally (and I'm pretty sure it works this way already but I'm not 100% certain) if the song is already downloaded it should play the downloaded copy rather than playing it from the server.

Yeah, this is the default. You can see whether or not the song is streaming or playing from a download on the bottom left of the player screen.

**Comment by Ge082** [comment]
Does the app do some kind of caching when a music has been played like Plexamp do ? That's maybe how they handle this ?

**Comment by rigrig** [comment]
> how to handle whether or not to automatically switch

I don't know about iOS, but IIRC on Android at least you can
a) request the current network state
b) get an event when the network changes

What I would like it to do:
- Add a setting to enable Automatic Offline Mode
- If the setting is enabled:
- When starting up, or the network state changes: 
   * if the network is connected and metered: go offline
   * if the network is connected and unmetered: go online
   * otherwise: do nothing.

That way users can still turn it on/off manually, and (briefly) losing connectivity won't cause surprises.

(I guess later on we might also like a setting for a "Home" network)

**Comment by flloschy** [comment]
Finally working on it :) 
https://github.com/jmshrv/finamp/pull/1139

**Comment by provokateurin** [comment]
I appreciated the improvements @flloschy, but I personally think the best way would be to get rid of offline mode.
It should be totally transparent to the user, whether they are online or not. With proper caching, there is no need to go into an offline mode.

@Chaphasilor does that make sense to you? If so, I will open a follow-up issue about this, so it can be implemented at some point.

**Comment by flloschy** [comment]
I would like to disagree 

Offline mode is important, not only for piece of mind but also to make the app faster and to show the user what they can actually listen to.

Piece of Mind: i always worry about cellular data usage and its nice to know that now im completely offline. Also people with metered connection or slow Internet probably appriciate offline mode.

Faster: Finamp always tries to fetch the newest data, with offline mode this is skipped which makes it more snappy.

Downloads: In offline mode only downloaded tracks are show, so when you Start listening there will never be a track which needs to be streamed while offline.


Regarding your transparency concerns, youll have multiple Modes for the automation: Off, No network connection, No Internet connetion. In the music player there is already an display to tell you if the song is player locally or streamed. And finally the offline button updates when offline mode is activated which the user can overwrite! 
But I like the idea of having a notification popping up, ill add that :)

**Comment by provokateurin** [comment]
> Piece of Mind: i always worry about cellular data usage and its nice to know that now im completely offline. Also people with metered connection or slow Internet probably appriciate offline mode.

But this applies to any app really, so I don't see why Finamp should have it while no other app I use or know has such a mode.

> Faster: Finamp always tries to fetch the newest data, with offline mode this is skipped which makes it more snappy. (ig this could be counted as a Problem it self but im fine with these loding times tbh)

Well first the cached data is displayed and then once the new data arrived the UI is updated. This would still be very snappy while allowing the user to get the updated data as well.

> Downloads: In offline mode only downloaded tracks are show, so when you Start listening there will never be a track which needs to be streamed while offline.

I agree, that this makes a lot of sense, but I'd rather see an option in the settings that only allows streaming on unmetered wifi and prevents in on cellular connection or metered wifi.

> Regarding your transparency concerns

This was not really about the popup or anything else, it was just me saying that the users should not have to manually deal with offline mode ever by simply doing caching.

**Comment by flloschy** [comment]
> But this applies to any app really, so I don't see why Finamp should have it while no other app I use or know has such a mode.

I count this as a + for finamp, like I said I like to know and have control over when or when not data is being downloaded/uploaded.

> Well first the cached data is displayed and then once the new data arrived the UI is updated. This would still be very snappy while allowing the user to get the updated data as well.

Im not against caching, it would indeed be very nice though the cache needs to also update at some point and an offline mode would still be useful to prevent such updates from happening if not welcome.

> I agree, that this makes a lot of sense, but I'd rather see an option in the settings that only allows streaming on unmetered wifi and prevents in on cellular connection or metered wifi.

That sounds to me just like a hidden automated offline mode at this point xD

> This was not really about the popup or anything else, it was just me saying that the users should not have to manually deal with offline mode ever by simply doing caching.

Excluding the caching part, thats what auto-offline effectively fixes, no?

**Comment by Chaphasilor** [comment]
@provokateurin so you're saying we should just hide the offline setting and switch automatically?  
Because offline mode + automatic switching really is what you seem to be suggesting. Your "prevent streaming on cellular" option would then be the automatic switch set to WiFi only.

I'm not a big fan of caching metadata. It leads to inconsistencies between server anf client, and causes UI jerking when new data arrives

Can you point me to a popular media player that doesn't use an offline mode but supports offline playback? Would be curious to see how it's done there!

So I definitely agree with @flloschy here :)

**Comment by provokateurin** [comment]
> I'm not a big fan of caching metadata. It leads to inconsistencies between server anf client, and causes UI jerking when new data arrives

I agree it can look bad, but given that libraries usually don't have huge changes every time you load them, it's not really going to be so bad in the end.

> Can you point me to a popular media player that doesn't use an offline mode but supports offline playback? Would be curious to see how it's done there!

For example Spotify just does it. You can still browse everything, even things you haven't downloaded. It just tells you that you are offline and you can not download or stream things, but that's about it.

---

One problem I see with only relying on connectivity checks are cases where the network is available but unstable or the server has some kind of problem. To properly detect those, you'd need to make the real request and then depending on the response use the cached data or update the data.
This is something not covered by the PR from @flloschy and I doesn't really align with the overall idea of a dedicated offline mode, since a server failure might only be temporary.

**Comment by Chaphasilor** [comment]
Amount of changes depends on the sorting. Recently added and last played change quite often.

Spotify's caching doesn't work well for me. Home screen is loaded fine, but going into the playlists and albums on the home screen only shows something half the time, for the other half I just get a spinner. For some other items I get a banner saying I'm offline and the data can't be loaded. Doesn't seem seamless to me 🤔

There's another proposal for your connection error point in #620, which I think is a reasonable request, and could work well with the automatic offline mode. That would just take 10s + one click, and then we could periodically retry the connection and show a banner prompting to leave offline mode.

**Comment by Chaphasilor** [comment]
@provokateurin oh and also, the original feature request by OP is about preventing mobile data usage. So we definitely need a toggle here that disables all network communication, which is exactly what offline mode is for :)

**Comment by Chaphasilor** [comment]
This has now been resolved as part of [beta version 0.9.18](https://github.com/jmshrv/finamp/releases/tag/0.9.18-beta)!

(was already a part of 0.9.16)
